types/encoding.go

```diff
@@ -170,12 +181,12 @@ func (d *Decoder) ReadUint64() uint64 {
 }
 
 // ReadPrefix reads a length prefix from the underlying stream. If the length
-// exceeds the number of bytes remaining in the stream, ReadPrefix sets d.Err
+// exceeds the maximum number of elements, ReadPrefix sets d.Err
 // and returns 0.
-func (d *Decoder) ReadPrefix() int {
+func (d *Decoder) ReadPrefix(limit int64) int {
```

Comment:
All of these should be reviewed carefully to ensure we're not going to break existing clients.

Reply:
Ahhh I see. That is quite unfortunate^^

Reply:
Is it that crazy to not pre-allocate the slice but grow it using `append`? Attackers could abuse the `limit` field and pass the maximum allowed for every type or situation right?

Reply:
append wouldn’t help since the decoder doesn’t stop decoding. It returns the zero value for anything after the error